coderd/schedule/autostop.go

```diff
@@ -0,0 +1,314 @@
+package schedule
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/coderd/database"
+)
+
+const (
+	// restartRequirementLeeway is the duration of time before a restart
+	// requirement where we skip the requirement and fall back to the next
+	// scheduled restart. This avoids workspaces being restarted too soon.
+	restartRequirementLeeway = 1 * time.Hour
+
+	// restartRequirementBuffer is the duration of time we subtract from the
+	// time when calculating the next scheduled restart time. This avoids issues
+	// where autostart happens on the hour and the scheduled quiet hours are
+	// also on the hour.
+	restartRequirementBuffer = -15 * time.Minute
```

Comment:
I think I get this, but commenting some examples might help illustrate

Reply:
I added examples on how these times help to each of them, let me know if I need to make them better